Add code for lambda lifting the backward in JIT's AD

apaszke opened the pull request:
Added the parts that were missing from yesterday's PR (namely generating not only `df`, but `f` graph + metadata to connect their outputs).

This PR still doesn't support having selective requires_grad, and assumes that all outputs of the first stage are meant to be differentiated. I'll add this to both `differentiate` and `lambdaLiftReverse` in a next patch.

apaszke committed:
Add code for lambda lifting backward in JIT's AD

yf225 commented:
@pytorchbot retest this please

apaszke commented:
@pytorchbot retest this please

zdevito commented:
I think we need to sync up before merging this. The specific requirements are subtle, and I want to make sure we are the same page. The pseudocode already has the meta-data structures that it will need to run the gradient, and I'd like to not introduce more.

zdevito commented on torch/csrc/jit/autodiff.cpp:
```diff
@@ -59,4 +57,170 @@ void differentiate(std::shared_ptr<Graph>& graph) {
+  // First, we need to find all values that are produced in the first stage,
+  // and used in the second one. They will need to be added as inputs of the reverse
+  // graph, and some of them may also need to be appended as outputs of the primal graph.
+  std::unordered_set<Value*> extra_reverse_inputs;
```
Do this the opposite way: look for definitions of stage 0 values used in stage one. This will naturally find them in topological order, no need for a separate pass.

zdevito commented on torch/csrc/jit/autodiff.cpp:
```diff
@@ -59,4 +57,170 @@ void differentiate(std::shared_ptr<Graph>& graph) {
+  // but artificially boosting outputs of the first stage to precede all
+  // other values.
+  auto topo_order = computeValueTopoOrder(graph);
+  auto io_compare = [&](Value *a, Value *b) { return topo_order.at(a) < topo_order.at(b); };
```
It's too confusing to encode the ordering of the outputs as a sort. I can't glance at this and know if it is correct. It would be easier if the temporaries were added to the primal as part of the process of splitting the graph rather than trying to construct one giant graph here.

apaszke replied:
I disagree. There is a certain ordering on the outputs we expect to get for both primal outputs and reverse inputs, and this is a simple way of enforcing them *consistently*, in a way such that `splitOffGraph` can remain a non-specialized helper, that doesn't really care if it's splitting the primal or dual. Moving that logic there would require adding a bunch of special cases, and it would certainly be harder for me to verify correctness. Here, you have a guarantee that if only this order is constructed correctly, then the outputs will look exactly as we would like them to.

zdevito commented on torch/csrc/jit/autodiff.cpp:
```diff
@@ -59,4 +57,170 @@ void differentiate(std::shared_ptr<Graph>& graph) {
+  value_list reverse_outputs = filter(graph->outputs(), is_stage_1);
+
+  // Cache some information that will be useful in stage 4.
+  auto num_original_primal_outputs = primal_outputs.size();
```
Too many random numbers used to slice arrays here.

apaszke replied:
Random numbers? I'm only using the number of inputs/outputs, and this code doesn't slice anything.

zdevito commented on torch/csrc/jit/autodiff.h:
```diff
@@ -12,4 +12,21 @@ namespace torch { namespace jit {
 void differentiate(std::shared_ptr<Graph>& graph);
 
+using value_list = std::vector<Value*>;
+struct LiftedReverse {
```
Please make this match the Gradient structure in the pseudo code.

apaszke replied:
This is not `Gradient` - if you look at the pseudo code, you'll see that half of its fields come from `differentiate`, while the other half come from `lambdaLiftReverse`. This lets us encapsulate the AD code in a fairly independent way, that doesn't need to know anything about `GraphExecutor`s.

zdevito commented on torch/csrc/jit/autodiff.cpp:
```diff
@@ -59,4 +57,170 @@ void differentiate(std::shared_ptr<Graph>& graph) {
+  //   [original outputs], [temporaries]
+  //
+  // Reverse inputs:
+  //   [original outputs#], [temporaries], [output vjps (aka grad_outputs)]
```
We need `grad_outputs` for the `[temporaries]` list as well, and the gradient calculation needs to be updated to add those values into the grad_temp variable in the reverse. This likely means that you will need to still have the `grad_map` around when you do this, and that this code need to be reorganized to make that possible.

apaszke replied:
Good point. Will need to update that 😕

apaszke commented:
One of the reasons why I wrote it this way, is exactly because the pseudo code is quite large by itself, and I didn't see a reason to shoehorn all of that AD stuff in the middle of it. Also, I think it's nice to keep the separation between `differentiate` and `lambdaLiftReverse` - I think this is our 3rd iteration of the JIT, so I wouldn't like to overspecialize to it right from the start. This way of writing the code makes it easier to plug in to other components, which might be needed in the future.

ezyang commented on torch/csrc/jit/test_jit.cpp:
```diff
@@ -536,8 +536,79 @@ void testADFormulas() {
+  graph->registerOutput(c.value());
+
+  static const char *expected_diff =
+R"(graph(%0 : Float(2, 3, 4)
```
As your local testing czar, BOOOOOOOOOO. (Make `--accept` work somehow, don't make me manually fix each of these diffs the next time we have a format change.)

ezyang commented on torch/csrc/jit/autodiff.cpp:
```diff
@@ -59,4 +57,170 @@ void differentiate(std::shared_ptr<Graph>& graph) {
+  return order;
+}
+
+LiftedReverse lambdaLiftReverse(std::shared_ptr<Graph>& graph) {
```
This function has implementation comments (good!), but even better would be a top-level comment saying, semantically, what this is supposed to do!

ezyang commented on torch/csrc/jit/autodiff.h:
```diff
@@ -12,4 +12,21 @@ namespace torch { namespace jit {
+  // - df_input_captures describes which outputs of f to capture and later provide
+  //   as first df_input_captures.size() inputs to df (with trailing inputs being
+  //   vjps aka grad outputs).
+  std::size_t f_output_intermediates;
```
aka `num_temporary_outputs`, in the pseudocode

ezyang commented on torch/csrc/jit/autodiff.h:
```diff
@@ -12,4 +12,21 @@ namespace torch { namespace jit {
+  //   as first df_input_captures.size() inputs to df (with trailing inputs being
+  //   vjps aka grad outputs).
+  std::size_t f_output_intermediates;
+  std::vector<std::size_t> df_input_captures;
```
aka `captured_inputs`, in the pseudocode. But I am a bit confused; don't you also need input captures? Unless you are making `f` forward those through but then that needs to be remarked upon here.

apaszke commented:
I think the problem in here is that we don't have nested declarations, and we're effectively splitting a 2-stage generator/coroutine into two lambdas - one that computes the results of the first stage *and returns the state*, and the other one that *consumes the state* also computes the results  (modulo the fact that the state implies adding more inputs to the second lambda, which makes it even worse). Our graph printing is very misleading and makes them look like regular functions whereas they really should look like this:
```
graph (%1) {
  %3 = add(%1, %1)
  %4 = yield (%3) -- NOTE: %4 is grad of %3
  --- stage 1 ---
  %5 = add(%4, %4)
  return (%5)
}
```
I don't think we can take a lot form the way it's implemented in functional languages, both because the objects we're dealing with are different, and because of how our IR is designed (e.g. it's much easier to clone the relevant parts with inputs added ahead of time instead of adding/remapping stuff inside the graph later).

ezyang commented:
> I think the problem in here is that we don't have nested declarations

But we are going to add this, because we need it for control flow, right?

> we're effectively splitting a 2-stage generator/coroutine into two lambdas - one that computes the results of the first stage and returns the state, and the other one that consumes the state also computes the results (modulo the fact that the state implies adding more inputs to the second lambda, which makes it even worse).

Yes. All I am saying is this can easily be modeled in a higher order language, because the first stage can just return a lambda which closes over the state that needs to be passed over. Then, if you don't want to actually execute a higher order language, you can lower this into explicit state passing.

apaszke committed:
Inherit JIT scopes when cloning only when it's correct

It's correct only when the new graph owns the same scope tree
as the original one. We can end up with dangling pointers otherwise.

zdevito commented on torch/csrc/jit/autodiff.h:
```diff
@@ -7,10 +7,59 @@
+  std::shared_ptr<Graph> f;
+  std::shared_ptr<Graph> df;
+
+  // Describes how to construct outputs of f from what its graph will return.
```
These comments are good -- can you also copy the example from the pseudo-code into here as well?

apaszke replied:
Hmm that will create a lot of duplication with the comment you requested. I'd rather put it in a single place, so I'll put it here as a note, and mention it over there.

zdevito reviewed (approved):
This looks good! I just have requests for more comments to clarify what each function is doing.

apaszke commented on torch/csrc/jit/autodiff.cpp:
```diff
@@ -34,38 +45,227 @@ static std::vector<Value*> gradientForNode(Node* node, ArrayRef<Value*> grad_val
+    graph.registerOutput(grad_map.at(input));
+    grad_desc.df_output_vjps.push_back(i);
+  }
+
```
This comment exists, but it's in `differentiate`. I'll repeat it here as well if you want

apaszke commented on torch/csrc/jit/autodiff.cpp:
```diff
@@ -34,38 +45,227 @@ static std::vector<Value*> gradientForNode(Node* node, ArrayRef<Value*> grad_val
+  // and used in the second one. They will need to be added as inputs of the reverse
+  // graph, and some of them may also need to be appended as outputs of the primal graph.
+  value_set reverse_captures_set;
+  value_list reverse_captures; // Invariant: topo sorted
```
Eh I'm using reverse throughout this function and I'd rather keep it consistent. Can I leave it as is?

ezyang commented on test/expect/TestJit.test_cpp.expect:
```diff
@@ -27,3 +27,25 @@ with GraphExecutor_0 = graph(%1 : UNKNOWN_TYPE
+}
+graph(%0 : Float(2, 3, 4)
+      %1 : Float(2, 3, 4)
+      %2 : Float(2, 3, 4)
```
At times like this I really wish we encoded the derivative metadata information into the names themselves; even something as simple as `d0` :)

apaszke replied:
Yes, I've been planning to add this at some point too

apaszke committed:
Fix a bug in BatchMM JIT pass

add node has multiple overloads, including one that only takes a
single input. This wasn't checked previously and could lead to
segfaults.

ezyang commented on torch/csrc/jit/autodiff.cpp:
```diff
@@ -34,38 +45,233 @@ static std::vector<Value*> gradientForNode(Node* node, ArrayRef<Value*> grad_val
+//   - graph has only stage 0
+//   - grad_desc doesn't have any fields initialized
+// After:
+//   - graph has stage 0 and stage 1 that computes its vjp
```
we should have a glossary somewhere for acronyms like "vjp" :)

ezyang commented on torch/csrc/jit/autodiff.cpp:
```diff
@@ -34,38 +45,233 @@ static std::vector<Value*> gradientForNode(Node* node, ArrayRef<Value*> grad_val
+//   - grad_desc has df_input_vjps and df_output_vjps set
+//     (but df_input_vjps will be modified later as well)
+static value_map addReverseInline(Graph& graph, Gradient& grad_desc) {
+  JIT_ASSERT(graph.stage() == 0);
```
Is stage() == 0 a fundamental limitation or simply "we haven't tested with stage() == 1, so dunno if it works or not"?

apaszke replied:
it's stage() == 0, because that's the regime it will be used in right now, and I don't want to generalize the semantics to arbitrary stages, because that will only make the code more complicated.

ezyang commented on torch/csrc/jit/autodiff.h:
```diff
@@ -7,10 +7,80 @@
+  // VJPs are "seeds" for the gradient computation given for each input capture
+  // of an Output kind.
+  std::vector<Capture> df_input_captures;
+  std::vector<std::size_t> df_input_vjps; // Offsets into f's outputs.
```
I'm not a fan of the member name `df_input_vjps` because no where does it cue that this is actually a vector of *indexes* into the outputs array (similar with `df_output_vjps`).

apaszke replied:
I'm not a huge fan, but it's ok. It tells you vjps for which outputs are inputs to df. I can't find a better name

ezyang commented on torch/csrc/jit/autodiff.cpp:
```diff
@@ -34,38 +45,233 @@ static std::vector<Value*> gradientForNode(Node* node, ArrayRef<Value*> grad_val
+    Value * output = outputs[i];
+    Value * output_grad = graph.addInput()->setType(output->typeOption());
+    set_grad(output, output_grad);
+    grad_desc.df_input_vjps.push_back(i);
```
So, this seems to imply that for i < num_outputs, grad_desc.df_input_vjps[i] == i, is that right? Will we ever reorder this?

apaszke replied:
Yes, that's right and it will be like this until we add `requires_grad` (when `df_input_vjps` will be a subset of this interval).

ezyang commented on torch/csrc/jit/autodiff.cpp:
```diff
@@ -34,38 +45,233 @@ static std::vector<Value*> gradientForNode(Node* node, ArrayRef<Value*> grad_val
+// compute all values in a given stage. An exception will be thrown if this is
+// not the case.
+static std::shared_ptr<Graph> splitOffStage(
+        Graph& graph,
```
const?

apaszke replied:
I remember there were some problems with const correctness. I think `cloneFrom` doesn't have a `const` overload, and I didn't want to pile more fixes onto this PR so I left this as is.

ezyang commented on torch/csrc/jit/autodiff.cpp:
```diff
@@ -34,38 +45,233 @@ static std::vector<Value*> gradientForNode(Node* node, ArrayRef<Value*> grad_val
+//   - values corresponding to outputs will be returned from the new graph
+// It requires that values contained in inputs are sufficient to be able to
+// compute all values in a given stage. An exception will be thrown if this is
+// not the case.
```
You also require outputs to be computed by the stage, right?

apaszke replied:
Well yes but that doesn't feel like it's unexpected.

ezyang commented on torch/csrc/jit/autodiff.cpp:
```diff
@@ -34,38 +45,233 @@ static std::vector<Value*> gradientForNode(Node* node, ArrayRef<Value*> grad_val
+
+// This function will take the graph and return a new one that:
+//   - contains all nodes of graph that have given stage
+//   - there will be an input corresponding to each input of the inputs array
```
I suppose it would clarify if you mentioned that an 'input' doesn't necessarily have to be an input of the graph itself; it may be a temporary from a previous stage you want to flow in.

ezyang commented on torch/csrc/jit/autodiff.cpp:
```diff
@@ -34,38 +45,233 @@ static std::vector<Value*> gradientForNode(Node* node, ArrayRef<Value*> grad_val
+  const auto lookup_val = [&](Value *v) { return val_map.at(v); };
+
+  for (Value *input : inputs)
+    val_map[input] = graph_clone->addInput()->setType(input->typeOption());
```
Opportunity for additional safety here: assert that the stage of input is <= stage.
